stDIKO as protocol token

nieldeckx opened the pull request:
- stDIKO is now minted/burned through the DAO
- DAO keeps track of contracts which can mint and/or burn

nieldeckx commented:
Stake-registry in pools is now dynamic. 

**Scenario 1 - Deactivate pool**
By pushing a new stake-registry we can replace a pool (deactivate old one on block X, activate new one on block X+1). The pool can still mint/burn tokens so stakers can use the 'unstake' method to retrieve funds and claim rewards. In the new stake-registry we can set when an old pool is deactivated, so reward emissions will stop at that block.

**Scenario 2 - Shut down pool**
Through the DAO we can decide which pools can mint and/or burn. If there is a critical bug we can prevent a pool from minting new protocol tokens. In this case the user can't use the 'unstake' as this method claims rewards, but the 'emergency-withdraw' method can be used to get back staked funds. The 'emergency-withdraw' method on the DIKO pool will burn stDIKO so although the pool can't mint anymore, it can still burn.

philipdesmedt commented on clarity/contracts/arkadiko-stake-registry-trait-v1.clar:
```diff
@@ -0,0 +1,9 @@
+(define-trait stake-registry-trait
```
Q: should we add more methods to this trait? If we need to do another dynamic `contract-call?`, it has to be in the trait.

nieldeckx replied:
A: Not sure to understand, which methods would you like to see in the trait as well?

philipdesmedt commented on clarity/contracts/arkadiko-stake-registry-v1-1.clar:
```diff
@@ -88,26 +107,10 @@
-      
-      (ok amount)
-    )
+    (asserts! (is-eq (as-contract tx-sender) (contract-of registry-trait)) ERR-WRONG-REGISTRY)
```
Q: all of these methods that have this `asserts!` cannot be called externally from a random addr, right?

Q2: on line 93 we still have a completely commented out method, do we still need that?

nieldeckx replied:
A: correct, the parameter must be the current registry. Will add some tests.
A2: it's not really needed as you can call the method on the pool directly. But it's nice to have all user interactions done via the registry, so will make sure to add the method and tests.

nieldeckx replied:
Both addressed in last commit.

philipdesmedt reviewed (changes requested):
Looks great.. mostly some questions

nieldeckx commented on clarity/contracts/arkadiko-stake-pool-diko-v1-1.clar:
```diff
@@ -120,30 +72,30 @@
 
 ;; Stake tokens
-(define-public (stake (token <ft-trait>) (staker principal) (amount uint))
+(define-public (stake (registry-trait <stake-registry-trait>) (token <ft-trait>) (staker principal) (amount uint))
```
The method can only be called by the stake-registry - see assert on L117
In the stake-registry itself we do check if the parameter is correct - see assert on L71

nieldeckx commented on clarity/contracts/arkadiko-stake-pool-diko-v1-1.clar:
```diff
@@ -157,7 +109,7 @@
 
 ;; Unstake tokens
-(define-public (unstake (token <ft-trait>) (staker principal) (amount uint))
+(define-public (unstake (registry-trait <stake-registry-trait>) (token <ft-trait>) (staker principal) (amount uint))
```
The method can only be called by the stake-registry - see assert on L117
In the stake-registry itself we do check if the parameter is correct - see assert on L71

philipdesmedt commented:
LGTM. I will test this first locally and merge after.

nieldeckx commented:
Done :) 

I renamed "arkadiko-stake-registry-v1-1_tests.ts" to "arkadiko-stake-pool-diko-xusd-v1-1_tests.ts" and changed all tests to work with the DIKO/xUSD pool. Added second test file for the DIKO pool.

Note that the "claim-pending-rewards" and "get-pending-rewards" methods are not relevant anymore for this pool, so you will need to update the FE.

nieldeckx commented:
Merged master and fixed conflicts.

philipdesmedt commented:
@nieldeckx cool - will review this asap

philipdesmedt commented on clarity/contracts/arkadiko-stake-pool-diko-xusd-v1-1.clar:
```diff
@@ -219,24 +222,25 @@
 
 ;; Calculate current cumm reward per stake
-(define-read-only (calculate-cumm-reward-per-stake)
+(define-public (calculate-cumm-reward-per-stake (registry-trait <stake-registry-trait>))
```
Q: shouldn't we check on correct stake registry here as well? A user can just call this with a random trait?

nieldeckx replied:
A: the method does not update any variables, so not critical to check. Can add it.

philipdesmedt replied:
Ok - no worries then. No need to add it. Just additional read cycles then.

philipdesmedt commented on clarity/contracts/xusd-token.clar:
```diff
@@ -81,8 +81,9 @@
   (try! (ft-mint? xusd u1000000000000 'STSTW15D618BSZQB85R058DS46THH86YQQY6XCB7)) ;; 1 million xUSD
   (try! (ft-mint? xusd u1000000000000 'ST1QV6WVNED49CR34E58CRGA0V58X281FAS1TFBWF)) ;; 1 million xUSD
+  (try! (ft-mint? xusd u1000000000000 'ST3EQ88S02BXXD0T5ZVT3KW947CRMQ1C6DMQY8H19)) ;; 1 million xUSD
+  (try! (ft-mint? xusd u1000000000000 'ST3KCNDSWZSFZCC6BE4VA9AXWXC9KEB16FBTRK36T)) ;; 1 million xUSD  
```
1 million dollars!!!

philipdesmedt reviewed (changes requested):
One small REQ. Lots of code so probably missed some things.. Gonna adapt the front-end now.

philipdesmedt commented on clarity/contracts/arkadiko-stake-pool-diko-v1-1.clar:
```diff
@@ -7,320 +7,137 @@
-)
-
-;; Get variable total-staked
-(define-read-only (get-total-staked)
```
@nieldeckx this was removed and we call this in the front-end. Should we implement this method on `arkadiko-stake-pool-diko-v1-1` and just call the FT supply function of stDIKO?

nieldeckx replied:
@philipdesmedt I will again add this method and calculate the amount of DIKO you would get if you unstaked with all the stDIKO from your wallet.

As DIKO is now auto-compounding you will see the 'staked value' increasing over time. This also means there is no need to show 'pending rewards' on this pool.

philipdesmedt replied:
OK thanks! I already added the method but feel free to adjust/fix

nieldeckx replied:
Fixed it. The method will now return the amount of DIKO you will get when unstaking - so including compounded rewards.

philipdesmedt replied:
Great, thx!

philipdesmedt commented:
@nieldeckx fixed the front-end (see my last 2 commits).. but the pending rewards always stay 0.. thoughts?
![image](https://user-images.githubusercontent.com/577420/123792303-c4258800-d8e0-11eb-81eb-c388979465af.png)

![image](https://user-images.githubusercontent.com/577420/123792327-c982d280-d8e0-11eb-8fcf-9680651bfdd4.png)

philipdesmedt commented:
@nieldeckx looking at the merge conflicts, maybe I shouldn't've merged the USDA rename PR 😂 

can you look if you can fix the merge conflicts easily?

nieldeckx commented:
@philipdesmedt fixed :)
